pkg/reloader/reloader.go

```diff
@@ -0,0 +1,213 @@
+// Copyright © 2023 Cisco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reloader
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/bank-vaults/vault-sdk/vault"
+	vaultapi "github.com/hashicorp/vault/api"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+func (c *Controller) runReloader(ctx context.Context) {
+	reloaderLogger := c.logger.WithContext(ctx).WithField("worker", "reloader")
+	reloaderLogger.Info("Reloader started")
+
+	if len(c.workloadSecrets.GetWorkloadSecretsMap()) == 0 {
+		reloaderLogger.Info("No secrets to reload")
+		return
+	}
+	if c.vaultClient == nil {
+		reloaderLogger.Info("Vault client is not initialized")
+		c.setVaultClient()
+	}
+
+	// Create a secretWorkloads map and compare the currently used secrets' version
+	// with the one stored in the secretVersions map
+	reloadedWorkloads := make(map[Workload]bool)
+	newSecretVersions := make(map[string]int)
+	for secretPath, workloads := range c.workloadSecrets.GetSecretWorkloadsMap() {
+		reloaderLogger.Debug("Checking secret: ", secretPath)
+		// Get current secret version
+		currentVersion, err := getSecretVersionFromVault(c.vaultClient, secretPath)
+		if err != nil {
+			if err, ok := err.(ErrSecretNotFound); ok {
+				reloaderLogger.Error(err.Error())
+				continue
+			}
+			c.vaultClient = nil
+			reloaderLogger.Error("failed to get secret version from vault")
+			utilruntime.HandleError(err)
+			return
+		}
+
+		// Compare current version with the secretVersions map
+		if c.secretVersions[secretPath] == 0 {
+			reloaderLogger.Debugf("Secret %s not found in secretVersions map, creating it", secretPath)
+			newSecretVersions[secretPath] = currentVersion
+			continue
+		}
+		if c.secretVersions[secretPath] == currentVersion {
+			reloaderLogger.Debugf("Secret %s did not change", secretPath)
+			newSecretVersions[secretPath] = currentVersion
+			continue
+		}
+		reloaderLogger.Debugf("Secret version stored: %d current: %d", c.secretVersions[secretPath], currentVersion)
+		for _, workload := range workloads {
+			if !reloadedWorkloads[workload] {
+				reloaderLogger.Info("Reloading workload: ", workload)
+				err := c.reloadWorkload(workload)
+				if err != nil {
+					utilruntime.HandleError(err)
+				}
+				reloadedWorkloads[workload] = true
+			}
+		}
```

Comment:
Instead of keeping track of the workloads that have been reloaded (and skipping if they are found in it), it would make more sense to have a map of workloads that need to be reloaded.
This way we separate fetching and processing from each other, making it more extendible.

```golang
workloadsToRefresh := make(map[Workload]bool)

// Fetching
for secretPath, workloads := range c.workloadSecrets.GetSecretWorkloadsMap() {
	// some code above
	for _, workload := range workloads {
		workloadsToRefresh[workload] = true
	}
}

// Processing (you can now process these parallely)
for workload := range workloadsToRefresh {
	reloaderLogger.Info("Reloading workload: ", workload)
	err := c.reloadWorkload(workload)
	if err != nil {
		utilruntime.HandleError(err)
	}
}

```

Reply:
This reads much better, thanks for the suggestion!